art/steps.py

```diff
@@ -833,3 +830,73 @@ def change_lr(model):
             model.lr = self.fine_tune_lr
 
         self.model_modifiers.append(change_lr)
+
+
+class Ensemble(ModelStep):
+    """This step tries to ensemble models"""
+
+    name = "Ensemble"
+    description = "Ensembles models"
+
+    def __init__(
+        self,
+        model: ArtModule,
+        num_models: int = 5,
+        logger: Optional[Logger] = None,
+        trainer_kwargs: Dict = {},
+        model_kwargs: Dict = {},
+        model_modifiers: List[Callable] = [],
+        datamodule_modifiers: List[Callable] = [],
+    ):
+        """
+        This method initializes the step
+
+        Args:
+            models (List[ArtModule]): models
+            logger (Logger, optional): logger. Defaults to None.
+            trainer_kwargs (Dict, optional): Kwargs passed to lightning Trainer. Defaults to {}.
+            model_kwargs (Dict, optional): Kwargs passed to model. Defaults to {}.
+            model_modifiers (List[Callable], optional): model modifiers. Defaults to [].
+            datamodule_modifiers (List[Callable], optional): datamodule modifiers. Defaults to [].
+        """
+        super().__init__(
+            model,
+            trainer_kwargs,
+            model_kwargs,
+            model_modifiers,
+            datamodule_modifiers,
+            logger=logger,
+        )
+        self.num_models = num_models
+
+    def do(self, previous_states: Dict):
+        """
+        This method trains the model
+
+        Args:
+            previous_states (Dict): previous states
+        """
+        models_paths = []
+        for _ in range(self.num_models):
+            self.reset_trainer(
+                logger=self.trainer.logger, trainer_kwargs=self.trainer_kwargs
+            )
+            self.train(trainer_kwargs={"datamodule": self.datamodule})
+            models_paths.append(self.trainer.checkpoint_callback.best_model_path)
+
+        initialized_models = []
+        for path in models_paths:
+            model = self.model_class.load_from_checkpoint(path)
+            model.eval()
+            initialized_models.append(model)
+
+        self.model = ArtEnsemble(initialized_models)
```

Comment:
I believe that self.model is never being used. Trainer is initialized and probably None is returned inside initialize model. So you in fact evaluate last model of the ensamble. Moreover this leads to a memory leak.

My suggestion is to add a flag `self.ensamble_ready` and overwrite `initialize_model` - If ensamble is ready it returns ensamble otherwise it calls super().initialize_model()